File: merge_sorting.cpp
#include <iostream>
#include <stdio.h>
#include <sys/time.h>
#include <ostream>
#include <fstream>

using namespace std;

namespace {
	ofstream take_log("data_merge_cpp.log");
}

int a[1000000];
int b[1000000];

double gettimeofday_sec(void)
{
	struct timeval t;
	gettimeofday(&t, NULL);
	return (double)t.tv_sec + (double)t.tv_usec * 1e-6;
}

void merge(int start, int end)
{
	int size = end - start + 1;
	int temp;
	int mid = (start + end) / 2;
	if (size > 2) {
		merge(start, mid);
		merge(mid + 1, end);
	}
	int one, two, top;
	one = start;
	two = mid + 1;
	top = start;

	while(1) {
		if (a[one] < a[two]) {
			b[top] = a[one];
			top++;
			one++;
			if (one > mid) {
				for (int i = two; i <= end; i++) {
					b[top] = a[i];
					top++;
				}
			}
		} else {
			b[top] = a[two];
			top++;
			two++;
			if (two > end) {
				for (int i = one; i <= mid; i++) {
					b[top] = a[i];
					top++;
				}
			}
		}
		if (top > end) break;
	}
	for (int i = start; i <= end; i++) {
		a[i] = b[i];
	}
}

int main(void)
{
	int size = 1;
	for (int t = 0; t < 18; t++) {
		size *= 2;
		for (int j = 0; j < size; j++) {
			a[j] = rand() % 1000;
		}

		double tstart, tend;
		tstart = gettimeofday_sec();
		merge(0, size - 1);
		tend = gettimeofday_sec();

		cout << tend - tstart << endl;
		take_log << tend - tstart << endl;
	}
	return 0;
}

File: merge_sorting.rb
def make1d(n)
	a = Array.new(n, 0)
	a
end

def merge(p, q)
	r = make1d(p.length() + q.length())
	first = 0
	second = 0
	for top in 0 .. r.length() - 1
		if first < p.length() && second < q.length()
			if p[first] < q[second]
				r[top] = p[first]
				first += 1
			else
				r[top] = q[second]
				second += 1
			end
		else 
			if first == p.length()
				r[top] = q[second]
				second += 1
			else 
				r[top] = p[first]
				first += 1
			end
		end
	end
	r
end			 

def randoms(id, size, max)
	a = Array.new(size)
	for i in 0.. size - 1
		if max == 1
			a[i]=rand()
		else
			a[i]=rand(max)
		end
	end
	a
end

def mergesort(s)
	mergesort_sub(s, 0, s.length())
end

def mergesort_sub(s, a, b)
	size = b - a
	if size > 1
		mid = (a + b) / 2
		p = mergesort_sub(s, a, mid)
		q = mergesort_sub(s, mid, b)
		r = merge(p, q)
	else
		r = make1d(1)
		r[0] = s[a]
	end
	r
end

def main(n)
	for i in 1 .. 18
		a = randoms(0, 2 ** i, 999)
		tstart = Time.now
		mergesort(a)
		tend = Time.now

		puts "#{tend - tstart}"
	end
end

